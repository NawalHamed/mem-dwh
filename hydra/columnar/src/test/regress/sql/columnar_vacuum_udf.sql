CREATE TABLE t1(a int, b int) USING columnar;

CREATE TABLE t2(a int, b int) USING columnar;

INSERT INTO t1 SELECT generate_series(1, 1000000, 1) AS a, generate_series(2, 2000000, 2) AS b;
INSERT INTO t2 SELECT generate_series(1, 2000000, 1) AS a, generate_series(2, 4000000, 2) AS b;

SELECT * FROM columnar.stats('t1'::regclass);
SELECT * FROM columnar.stats('t2'::regclass);

SELECT * FROM columnar.vacuum('t1'::regclass);

SELECT * FROM columnar.stats('t1'::regclass);
SELECT * FROM columnar.stats('t2'::regclass);

SELECT * FROM columnar.vacuum('t2'::regclass);

SELECT * FROM columnar.stats('t1'::regclass);
SELECT * FROM columnar.stats('t2'::regclass);

DELETE FROM t1 WHERE a % 2 = 0;
DELETE FROM t1 WHERE a % 100 = 0;

SELECT * FROM columnar.stats('t1'::regclass);
SELECT * FROM columnar.stats('t2'::regclass);

SELECT * FROM columnar.vacuum('t1'::regclass);
SELECT * FROM columnar.vacuum('t2'::regclass);

SELECT * FROM columnar.stats('t1'::regclass);
SELECT * FROM columnar.stats('t2'::regclass);

SELECT COUNT(*) FROM t1;
SELECT COUNT(*) FROM t2;

SELECT SUM(a), SUM(b) FROM t1;
SELECT SUM(a), SUM(b) FROM t2;

INSERT INTO t1 SELECT generate_series(1000001, 2000000, 1) AS a, generate_series(2000002, 4000000, 2) AS b;
-- should show no holes between the previous insert and this insert
SELECT * FROM columnar.stats('t1'::regclass);

DELETE FROM t1 WHERE a BETWEEN 500000 AND 1500000;

SELECT * FROM columnar.vacuum('t1'::regclass);
-- should show no holes
SELECT * FROM columnar.stats('t1'::regclass);

DROP TABLE t1;
DROP TABLE t2;

CREATE TABLE t1(a int, b int) USING columnar;
INSERT INTO t1 SELECT generate_series(1, 1000000, 1) AS a, generate_series(2, 2000000, 2) AS b;

DELETE FROM t1 WHERE a % 3 = 0;

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
SELECT * FROM columnar.vacuum('t1'::regclass, 1);

DROP TABLE t1;

CREATE table t1 (
  i1 int,
  i2 int,
  i3 int,
  i4 int
) using columnar;
INSERT INTO t1 SELECT t, t, t, t FROM generate_series(1, 7000000) t;
DELETE FROM t1 WHERE i1 % 13 = 0;
UPDATE t1 SET i1 = i1 * 2, i2 = i2 * 2, i3 = i3 * 3, i4 = i4 * 4;
UPDATE t1 SET i4 = null where i4 % 23 = 0;

SELECT columnar.vacuum('t1');

SELECT count(i1), count(i2), count(i3), count(i4) FROM t1;

DROP TABLE t1;

CREATE TABLE cache_test (i INT) USING columnar;
INSERT INTO cache_test SELECT generate_series(1, 100, 1);

SET columnar.enable_column_cache = 't';
VACUUM cache_test;

-- should be true
SHOW columnar.enable_column_cache;

ANALYZE cache_test;

-- should be true
SHOW columnar.enable_column_cache;

DROP TABLE cache_test;
