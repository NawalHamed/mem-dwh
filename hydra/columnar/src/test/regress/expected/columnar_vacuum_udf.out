CREATE TABLE t1(a int, b int) USING columnar;
CREATE TABLE t2(a int, b int) USING columnar;
INSERT INTO t1 SELECT generate_series(1, 1000000, 1) AS a, generate_series(2, 2000000, 2) AS b;
INSERT INTO t2 SELECT generate_series(1, 2000000, 1) AS a, generate_series(2, 4000000, 2) AS b;
SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   100000 |           0 |         10 |     543291
(7 rows)

SELECT * FROM columnar.stats('t2'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   150000 |           0 |         15 |     815582
        8 |    5709432 |   150000 |           0 |         15 |     812780
        9 |    6526232 |   150000 |           0 |         15 |     814376
       10 |    7343032 |   150000 |           0 |         15 |     814484
       11 |    8159832 |   150000 |           0 |         15 |     815362
       12 |    8976632 |   150000 |           0 |         15 |     813439
       13 |    9793432 |   150000 |           0 |         15 |     814608
       14 |   10610232 |    50000 |           0 |          5 |     272361
(14 rows)

SELECT * FROM columnar.vacuum('t1'::regclass);
 vacuum 
--------
      0
(1 row)

SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   100000 |           0 |         10 |     543291
(7 rows)

SELECT * FROM columnar.stats('t2'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   150000 |           0 |         15 |     815582
        8 |    5709432 |   150000 |           0 |         15 |     812780
        9 |    6526232 |   150000 |           0 |         15 |     814376
       10 |    7343032 |   150000 |           0 |         15 |     814484
       11 |    8159832 |   150000 |           0 |         15 |     815362
       12 |    8976632 |   150000 |           0 |         15 |     813439
       13 |    9793432 |   150000 |           0 |         15 |     814608
       14 |   10610232 |    50000 |           0 |          5 |     272361
(14 rows)

SELECT * FROM columnar.vacuum('t2'::regclass);
 vacuum 
--------
      0
(1 row)

SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   100000 |           0 |         10 |     543291
(7 rows)

SELECT * FROM columnar.stats('t2'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   150000 |           0 |         15 |     815582
        8 |    5709432 |   150000 |           0 |         15 |     812780
        9 |    6526232 |   150000 |           0 |         15 |     814376
       10 |    7343032 |   150000 |           0 |         15 |     814484
       11 |    8159832 |   150000 |           0 |         15 |     815362
       12 |    8976632 |   150000 |           0 |         15 |     813439
       13 |    9793432 |   150000 |           0 |         15 |     814608
       14 |   10610232 |    50000 |           0 |          5 |     272361
(14 rows)

DELETE FROM t1 WHERE a % 2 = 0;
DELETE FROM t1 WHERE a % 100 = 0;
SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |       75000 |         15 |     788928
        2 |     808632 |   150000 |       75000 |         15 |     815091
        3 |    1625432 |   150000 |       75000 |         15 |     814184
        4 |    2442232 |   150000 |       75000 |         15 |     813607
        5 |    3259032 |   150000 |       75000 |         15 |     813550
        6 |    4075832 |   150000 |       75000 |         15 |     815135
        7 |    4892632 |   100000 |       50000 |         10 |     543291
(7 rows)

SELECT * FROM columnar.stats('t2'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   150000 |           0 |         15 |     815582
        8 |    5709432 |   150000 |           0 |         15 |     812780
        9 |    6526232 |   150000 |           0 |         15 |     814376
       10 |    7343032 |   150000 |           0 |         15 |     814484
       11 |    8159832 |   150000 |           0 |         15 |     815362
       12 |    8976632 |   150000 |           0 |         15 |     813439
       13 |    9793432 |   150000 |           0 |         15 |     814608
       14 |   10610232 |    50000 |           0 |          5 |     272361
(14 rows)

SELECT * FROM columnar.vacuum('t1'::regclass);
 vacuum 
--------
     12
(1 row)

SELECT * FROM columnar.vacuum('t2'::regclass);
 vacuum 
--------
      0
(1 row)

SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        8 |      16336 |   150000 |           0 |         15 |     808748
        9 |     825084 |   150000 |           0 |         15 |     822409
       10 |    1647493 |   150000 |           0 |         15 |     821020
       11 |    2468513 |    50000 |           0 |          5 |     273902
(4 rows)

SELECT * FROM columnar.stats('t2'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        1 |      16336 |   150000 |           0 |         15 |     788928
        2 |     808632 |   150000 |           0 |         15 |     815091
        3 |    1625432 |   150000 |           0 |         15 |     814184
        4 |    2442232 |   150000 |           0 |         15 |     813607
        5 |    3259032 |   150000 |           0 |         15 |     813550
        6 |    4075832 |   150000 |           0 |         15 |     815135
        7 |    4892632 |   150000 |           0 |         15 |     815582
        8 |    5709432 |   150000 |           0 |         15 |     812780
        9 |    6526232 |   150000 |           0 |         15 |     814376
       10 |    7343032 |   150000 |           0 |         15 |     814484
       11 |    8159832 |   150000 |           0 |         15 |     815362
       12 |    8976632 |   150000 |           0 |         15 |     813439
       13 |    9793432 |   150000 |           0 |         15 |     814608
       14 |   10610232 |    50000 |           0 |          5 |     272361
(14 rows)

SELECT COUNT(*) FROM t1;
 count  
--------
 500000
(1 row)

SELECT COUNT(*) FROM t2;
  count  
---------
 2000000
(1 row)

SELECT SUM(a), SUM(b) FROM t1;
     sum      |     sum      
--------------+--------------
 250000000000 | 500000000000
(1 row)

SELECT SUM(a), SUM(b) FROM t2;
      sum      |      sum      
---------------+---------------
 2000001000000 | 4000002000000
(1 row)

INSERT INTO t1 SELECT generate_series(1000001, 2000000, 1) AS a, generate_series(2000002, 4000000, 2) AS b;
-- should show no holes between the previous insert and this insert
SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        8 |      16336 |   150000 |           0 |         15 |     808748
        9 |     825084 |   150000 |           0 |         15 |     822409
       10 |    1647493 |   150000 |           0 |         15 |     821020
       11 |    2468513 |    50000 |           0 |          5 |     273902
       12 |    2744448 |   150000 |           0 |         15 |     815176
       13 |    3561248 |   150000 |           0 |         15 |     813187
       14 |    4378048 |   150000 |           0 |         15 |     814516
       15 |    5194848 |   150000 |           0 |         15 |     814283
       16 |    6011648 |   150000 |           0 |         15 |     815497
       17 |    6828448 |   150000 |           0 |         15 |     813704
       18 |    7645248 |   100000 |           0 |         10 |     543338
(11 rows)

DELETE FROM t1 WHERE a BETWEEN 500000 AND 1500000;
SELECT * FROM columnar.vacuum('t1'::regclass);
 vacuum 
--------
     12
(1 row)

-- should show no holes
SELECT * FROM columnar.stats('t1'::regclass);
 stripeid | fileoffset | rowcount | deletedrows | chunkcount | datalength 
----------+------------+----------+-------------+------------+------------
        8 |      16336 |   150000 |           0 |         15 |     808748
       16 |     825084 |   150000 |           0 |         15 |     815497
       17 |    1640581 |   150000 |           0 |         15 |     813704
       18 |    2454285 |   100000 |           0 |         10 |     543338
       19 |    2997623 |   150000 |           0 |         15 |     819245
       20 |    3816868 |    50000 |           0 |          5 |     271619
(6 rows)

DROP TABLE t1;
DROP TABLE t2;
CREATE TABLE t1(a int, b int) USING columnar;
INSERT INTO t1 SELECT generate_series(1, 1000000, 1) AS a, generate_series(2, 2000000, 2) AS b;
DELETE FROM t1 WHERE a % 3 = 0;
SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      1
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      0
(1 row)

SELECT * FROM columnar.vacuum('t1'::regclass, 1);
 vacuum 
--------
      0
(1 row)

DROP TABLE t1;
CREATE table t1 (
  i1 int,
  i2 int,
  i3 int,
  i4 int
) using columnar;
INSERT INTO t1 SELECT t, t, t, t FROM generate_series(1, 7000000) t;
DELETE FROM t1 WHERE i1 % 13 = 0;
UPDATE t1 SET i1 = i1 * 2, i2 = i2 * 2, i3 = i3 * 3, i4 = i4 * 4;
UPDATE t1 SET i4 = null where i4 % 23 = 0;
SELECT columnar.vacuum('t1');
 vacuum 
--------
     93
(1 row)

SELECT count(i1), count(i2), count(i3), count(i4) FROM t1;
  count  |  count  |  count  |  count  
---------+---------+---------+---------
 6461539 | 6461539 | 6461539 | 6180603
(1 row)

DROP TABLE t1;
CREATE TABLE cache_test (i INT) USING columnar;
INSERT INTO cache_test SELECT generate_series(1, 100, 1);
SET columnar.enable_column_cache = 't';
VACUUM cache_test;
-- should be true
SHOW columnar.enable_column_cache;
 columnar.enable_column_cache 
------------------------------
 on
(1 row)

ANALYZE cache_test;
-- should be true
SHOW columnar.enable_column_cache;
 columnar.enable_column_cache 
------------------------------
 on
(1 row)

DROP TABLE cache_test;
